woocommerce/woocommerce-ios
# Retry server-side payment capture after client-side success

Closes: #6338 


### Description


As proposed in pdfdoF-Ip-p2, we can improve the error screen after unsuccessful server-side processing and gives merchants an option to retry without starting over the IPP flow (which we don't want for Interac payments, which Stripe will throw an error anyway). The merchant can still dismiss the alert to go back to the order, with a risk of the order being out of sync with Stripe for Interac payments and the payment not being captured for non-Interac payments.

I struggled a bit with mixed usage of completion blocks and Combine 😅 the major changes include:
- Yosemite layer: added a new action `CardPresentPaymentAction.capturePaymentOnSite` for server-side payment capture to allow retry
- App layer:
  - In `PaymentCaptureOrchestrator`, added a function `retryServerSidePaymentCapture` that dispatches `CardPresentPaymentAction.capturePaymentOnSite` and subscribes to the result of the API request. Upon success, it calls an internal function that is also called in a successful `collectPayment` flow
  - A new alert modal class was created `CardPresentModalServerSidePaymentCaptureError` to show the server-side payment capture error with a different retry handling. I first tried reusing `CardPresentModalError` but the logic grew quite complicated to show different messages and handle the primary action differently based on the error type
  - In `CollectOrderPaymentUseCase`, when receiving an error `ServerSidePaymentCaptureError.paymentGateway` it showed the new modal `CardPresentModalServerSidePaymentCaptureError` with a retry call to `PaymentCaptureOrchestrator.retryServerSidePaymentCapture`

### Testing instructions


We can simulate the server-side processing error manually by updating the `capture_terminal_payment` endpoint to an invalid path [here](https://href.li/?https://github.com/woocommerce/woocommerce-ios/blob/d91c96b72076fbe58990494a8c53b98c7dfdae88/Networking/Networking/Remote/WCPayRemote.swift#L30) (like adding any character at the end). Any other ways to generate an error after client-side processing like during receipt generation also work.

- Go to the orders tab
- Create an order and tap “Collect Payment” when it becomes available
- Connect to a reader and pay with an eligible card –> the reader should show the payment as successful
- Remove card –> the app should show a modal about the failure as in the "after" screenshot below (please feel free to suggest other messages in the alert!)
- Tap "Try Again" --> the app should try making the same API request with an error if you hard-coded it to always fail. Then, the app should show the same error modal again
- Tap “Back to Order” –> notice that “Collect Payment” CTA is still available

#### Retry success

- [x] @jaclync tested this

Feel free to skip this test case since it's trickier to simulate this scenario. I added a private variable to `WCPayRemote` to use the correct `capture_terminal_payment` endpoint when the count is even 😅 

- Go to the orders tab
- Create an order and tap “Collect Payment” when it becomes available
- Connect to a reader and pay with an eligible card –> the reader should show the payment as successful
- Remove card –> the app should show a modal about the failure as in the "after" screenshot below (please feel free to suggest other messages in the alert!)
- Tap "Try Again" --> the app should try making the same API request with success, then shows the success modal with actions to print/email receipt

### Screenshots


before | after
-- | --
![image](https://user-images.githubusercontent.com/1945542/169738625-c1d5994b-b15f-434c-915b-cf56c54edc67.png) | ![IMG_3468](https://user-images.githubusercontent.com/1945542/169739604-e9bf34a5-ef57-443f-9c33-8e0d176f91aa.PNG)


---
